aten/src/ATen/native/Math.h

```diff
@@ -1152,76 +1186,395 @@ scalar_t calc_igamma(scalar_t a, scalar_t x) {
+  using acc_t = at::acc_type<scalar_t, is_cuda>;
+  constexpr acc_t ZERO = acc_t{0.0};
+  constexpr acc_t ONE = acc_t{1.0};
+  constexpr acc_t SMALL = acc_t{20.0};
+  constexpr acc_t LARGE = acc_t{200.0};
+  constexpr acc_t SMALLRATIO = acc_t{0.3};
+  constexpr acc_t LARGERATIO = acc_t{4.5};
+  acc_t absxma_a;
 
   // boundary values following SciPy
   // note that in SciPy, a and x are non-negative, with exclusive 0s (i.e.,
   // at most 1 of them can be 0), where igamma(0, x) = 1.0 iff x > 0.
-  if ((x < 0) || (a < 0)) {
+  if ((x < ZERO) || (a < ZERO)) {
     // out of defined-region of the function
     return std::numeric_limits<scalar_t>::quiet_NaN();
   }
-  else if (a == 0) {
-    if (x > 0) {
-      return 1.0;
+  else if (a == ZERO) {
+    if (x > ZERO) {
+      return static_cast<scalar_t>(ONE);
     }
     else {
       return std::numeric_limits<scalar_t>::quiet_NaN();
     }
   }
-  else if (x == 0) {
-    return 0.0; // zero integration limit
+  else if (x == ZERO) {
+    return static_cast<scalar_t>(ZERO); // zero integration limit
   }
   else if (std::isinf(a)) {
     if (std::isinf(x)) {
       return std::numeric_limits<scalar_t>::quiet_NaN();
     }
-    return 0.0;
+    return static_cast<scalar_t>(ZERO);
   }
   else if (std::isinf(x)) {
-    return 1.0;
+    return static_cast<scalar_t>(ONE);
   }
 
   /* Asymptotic regime where a ~ x. See [igam2] */
-  absxma_a = std::fabs(x - a) / a;
+  absxma_a = ::fabs(x - a) / a;
   if ((a > SMALL) && (a < LARGE) && (absxma_a < SMALLRATIO)) {
-    return _igam_helper_asymptotic_series(a, x, 1);
+    return _igam_helper_asymptotic_series<scalar_t, is_cuda>(a, x, true);
   }
-  else if ((a > LARGE) && (absxma_a < LARGERATIO / std::sqrt(a))) {
-    return _igam_helper_asymptotic_series(a, x, 1);
+  else if ((a > LARGE) && (absxma_a < LARGERATIO / ::sqrt(a))) {
+    return _igam_helper_asymptotic_series<scalar_t, is_cuda>(a, x, true);
   }
 
-  if ((x > 1.0) && (x > a)) {
-    return 1.0 - calc_igammac(a, x);
+  if ((x > ONE) && (x > a)) {
+    return static_cast<scalar_t>(ONE - calc_igammac<scalar_t, is_cuda>(a, x));
   }
 
-  return _igam_helper_series(a, x);
+  return _igam_helper_series<scalar_t, is_cuda>(a, x);
+
+} // calc_igamma
+
+template <>
+C10_UNUSED C10_HOST_DEVICE inline c10::BFloat16
+calc_igamma<c10::BFloat16, false>(c10::BFloat16 a, c10::BFloat16 x) {
+  return calc_igamma<float, false>(float(a), float(x));
+}
+
+template <>
+C10_UNUSED C10_HOST_DEVICE inline c10::Half
+calc_igamma<c10::Half, false>(c10::Half a, c10::Half x) {
+  return calc_igamma<float, false>(float(a), float(x));
+}
+
+template <>
+C10_UNUSED C10_HOST_DEVICE inline c10::BFloat16
+calc_igammac<c10::BFloat16, false>(c10::BFloat16 a, c10::BFloat16 x) {
+  return calc_igammac<float, false>(float(a), float(x));
+}
+
+template <>
+C10_UNUSED C10_HOST_DEVICE inline c10::Half
+calc_igammac<c10::Half, false>(c10::Half a, c10::Half x) {
+  return calc_igammac<float, false>(float(a), float(x));
 }
```

Comment:
I know this is copy-pasted from existing code, but it would be  better to let opmath_type handle this automatically and drop the template specializations.